roadmap/implementers-guide/src/types/approval.md

```diff
@@ -0,0 +1,128 @@
+# Approval Types
+
+## ApprovalId
+
+The public key of a keypair used by a validator for approval voting.
+
+## AssignmentId
+
+The private key of a keypair used by a validator for approval voting.
+
+## AssignmentCert
+
+An `AssignmentCert`, short for Assignment Certificate, is a piece of data provided by a validator to prove that they have been selected to perform secondary approval checks on an included candidate.
+
+These certificates can be checked in the context of a specific block, candidate, and validator assignment VRF key. The block state will also provide further context about the availability core states at that block.
+
+```rust
+enum AssignmentCertKind {
+    RelayVRFModulo {
+        relay_vrf: (VRFInOut, VRFProof),
+        sample: u32,
+    },
+    RelayVRFDelay {
+        relay_vrf: (VRFInOut, VRFProof),
+        core_index: CoreIndex,
+    }
+}
+
+struct AssignmentCert {
+    // The criterion which is claimed to be met by this cert.
+    kind: AssignmentCertKind,
+    // The VRF showing the criterion is met.
+    vrf: VRFInOut,
+}
+```
+
+> TODO: RelayEquivocation cert. Probably can only be broadcast to chains that have handled an equivocation report.
+
+## ApprovalVote
+
+A vote of approval on a candidate.
+
+```rust
+struct ApprovalVote(Hash);
+```
+
+## SignedApprovalVote
+
+```rust
+struct SignedApprovalVote {
+    vote: ApprovalVote,
+    validator: ValidatorIndex,
+    signature: ApprovalSignature,
+}
+```
+
+## IndirectSignedApprovalVote
+
+A signed approval vote which references the candidate indirectly via the block. If there exists a look-up to the candidate hash from the block hash and candidate index, then this can be transformed into a `SignedApprovalVote`.
+
+Although this vote references the candidate by a specific block hash and candidate index, the vote actually applies to
+
+```rust
+struct IndirectSignedApprovalVote {
+    // A block hash where the candidate appears.
+    block_hash: Hash,
+    // The index of the candidate in the list of candidates fully included as-of the block.
+    candidate_index: u32,
+    validator: ValidatorIndex,
+    signature: ApprovalSignature,
+}
+```
```

Comment:
Yes, the signature is always over `ApprovalVote`, but we handle it with `block_hash` and `candidate_index` sometimes.  Would we gossip in the `SignedApprovalVote` form?  We could use an abbreviation of `IndirectSignedApprovalVote` when placing on-chain perhaps, although maybe not initially?.

Reply:
I'd expect we gossip in the `Indirect` form, because it allows us to gauge politeness more easily. That's the main purpose of this struct, actually.

Reply:
We don't have a good way of knowing if a peer has a candidate, except for knowing if they have the block that caused inclusion of the candidate.